path_selection/path_selection.go

```diff
@@ -0,0 +1,275 @@
+		return nil, err
+	}
+	descriptors[r.nrHops-1], err = r.pki.GetProviderDescriptor(recipientProviderName)
+	if err != nil {
+		return nil, err
+	}
+	for i := 0; i < r.nrHops; i++ {
+		layerMixes := r.pki.GetMixesInLayer(uint8(i))
+		fmt.Printf("layer %d\n", i)
+		if len(layerMixes) == 0 {
+			panic("WTF")
+		}
+		c := mathrand.Intn(len(layerMixes))
+		descriptors[i] = layerMixes[c]
+	}
+	return descriptors, nil
+}
+
+// getHopEpochKeys is a helper function which ultimately selects
+// appropriate mix routing keys for each hop in the route. This function is
+// given a 'till' argument which specifies the mount of time until the
+// next Key Rotation Epoch. The 'delays' argument specifies the delay for
+// each hop in the route and the 'descriptors' is a list of MixDescriptor
+// for each hop. For detailed information about the mix key rotation
+// schedule, refer to section "4.2 Sphinx Mix and Provider Key Rotation"
+// of the "Panoramix Mix Network Specification"
+// https://github.com/Katzenpost/docs/blob/master/specs/mixnet.txt
+func (r *RouteFactory) getHopEpochKeys(till time.Duration, delays []float64, descriptors []*pki.MixDescriptor) ([]*ecdh.PublicKey, error) {
+	hopDelay := delays[0]
+	keys := make([]*ecdh.PublicKey, r.nrHops)
+	for i := 0; i < len(descriptors); i++ {
+		hopDelay += delays[i]
+		hopDuration, err := durationFromFloat(hopDelay)
+		if err != nil {
+			return nil, err
+		}
+		if hopDuration < till {
+			keys[i] = descriptors[i].EpochAPublicKey
+		} else if hopDuration > till && hopDuration < till+epochtime.Period {
+			keys[i] = descriptors[i].EpochBPublicKey
+		} else if hopDuration > till && hopDuration < till+(2*epochtime.Period) {
+			keys[i] = descriptors[i].EpochCPublicKey
+		} else {
+			return nil, errors.New("error: inappropriate delays")
+		}
+	}
+	return keys, nil
+}
+
+// newPathVector returns a slice of PathHops and optionally a new SURB ID
+// if the isSURB argument is set to true.
+// The PathHop struct has three attributes: ID, PublicKey and Commands.
+// The ID and PublicKey are found in the mix descriptors while the
+// Commands are specified by this function. The recipientID is only
+// used to create forward paths, when isSURB is set to false.
+func (r *RouteFactory) newPathVector(till time.Duration,
+	delays []float64,
+	descriptors []*pki.MixDescriptor,
+	recipientID [constants.RecipientIDLength]byte,
+	isSURB bool) ([]*sphinx.PathHop, *[constants.SURBIDLength]byte, error) {
+
+	path := make([]*sphinx.PathHop, r.nrHops)
+	var surbID *[constants.SURBIDLength]byte = nil
+	keys, err := r.getHopEpochKeys(till, delays, descriptors)
+	if err != nil {
+		return nil, nil, err
+	}
+	for i := range path {
+		path[i] = new(sphinx.PathHop)
+		copy(path[i].ID[:], descriptors[i].ID[:])
+		path[i].PublicKey = keys[i]
+		if i < r.nrHops-1 {
+			// Non-terminal hop, add the delay.
+			delay := new(commands.NodeDelay)
+			delay.Delay = uint32(delays[i])
+			path[i].Commands = append(path[i].Commands, delay)
```

Comment:
what will cause `path[i].Commands` to already have commands? can't it be directly set to an army with the single element on lines 174, 184, and 189?

Reply:
yes that sounds good.